lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -953,47 +956,96 @@ static void deriveBodyDecodable_init(AbstractFunctionDecl *initDecl) {
+          CallExpr::createImplicit(C, superDecoderRef, ArrayRef<Expr *>(),
+                                   ArrayRef<Identifier>());
+
+        // super
+        auto *superRef = new (C) SuperRefExpr(initDecl->getImplicitSelfDecl(),
+                                              SourceLoc(), /*Implicit=*/true);
+
+        // super.init(from:)
+        auto initName = DeclName(C, C.Id_init, (Identifier[1]){ C.Id_from });
+        auto *initCall = new (C) UnresolvedDotExpr(superRef, SourceLoc(),
+                                                   initName, DeclNameLoc(),
+                                                   /*Implicit=*/true);
 
-    // try super.init(from: container.superDecoder())
-    auto *tryExpr = new (C) TryExpr(SourceLoc(), callExpr, Type(),
-                                    /*Implicit=*/true);
-    statements.push_back(tryExpr);
+        // super.decode(from: container.superDecoder())
+        Expr *args[1] = {superDecoderCall};
+        Identifier argLabels[1] = {C.Id_from};
+        auto *callExpr = CallExpr::createImplicit(C, initCall,
+                                                  C.AllocateCopy(args),
+                                                  C.AllocateCopy(argLabels));
+
+        // try super.init(from: container.superDecoder())
+        auto *tryExpr = new (C) TryExpr(SourceLoc(), callExpr, Type(),
+                                        /*Implicit=*/true);
+        statements.push_back(tryExpr);
+      } else {
+        // The explicit constructor name is a compound name taking no arguments.
+        DeclName initName(C, C.Id_init, ArrayRef<Identifier>());
+
+        // We need to look this up in the superclass to see if it throws.
+        // We should have bailed one level up if super.init() is not available.
+        ConstructorDecl *superInitDecl = nullptr;
+        for (auto *decl : superclassDecl->getMembers()) {
+          auto *ctorDecl = dyn_cast<ConstructorDecl>(decl);
+          if (!ctorDecl)
+            continue;
+
+          if (ctorDecl->getParameters()->size() == 0) {
+            superInitDecl = ctorDecl;
+            break;
+          }
+        }
+
+        // We should have bailed one level up if this were not available.
+        assert(superInitDecl);
+
+        // super
+        auto *superRef = new (C) SuperRefExpr(initDecl->getImplicitSelfDecl(),
+                                              SourceLoc(), /*Implicit=*/true);
+
+        // super.init()
+        auto *superInitRef = new (C) UnresolvedDotExpr(superRef, SourceLoc(),
+                                                       initName, DeclNameLoc(),
+                                                       /*Implicit=*/true);
+        // super.init() call
+        auto *callExpr = CallExpr::createImplicit(C, superInitRef,
+                                                  ArrayRef<Expr *>(),
+                                                  ArrayRef<Identifier>());
+
+        if (superInitDecl->hasThrows()) {
```

Comment:
Yeah, I think that's the only way. I'm just not sure how much you/we care about these cases at all.